### ISSUE TITLE
`GLIBCXX_3.4.20' not found

### ISSUE THREAD

**Issue by GA-Goig** [infphilo/centrifuge#79]
Hi, 

I'm getting the following error when executing centrifuge on CentOS

### My SO
```
dmesg | grep "Linux version"
[    0.000000] Linux version 3.10.0-327.22.2.el7.x86_64 (builder@kbuilder.dev.centos.org) (gcc version 4.8.3 20140911 (Red Hat 4.8.3-9) (GCC) ) #1 SMP Thu Jun 23 17:05:11 UTC 2016
```

### The problem
```
/usr/local/bin/centrifuge-class: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /usr/local/bin/centrifuge-class)
(ERR): Description of arguments failed!
Exiting now ...
```
Is the same error when calling just centrifuge or when passing arguments. I tried several things like set the LD_LIBRARY_PATH or download new libdstdc versions. I get the same problem either, when compiling from source or when running the prebuilt binary.

And I think is not a problem with the library itself

```
strings /lib64/libstdc++.so.6 | grep GLIBCXX
GLIBCXX_3.4
GLIBCXX_3.4.1
GLIBCXX_3.4.2
GLIBCXX_3.4.3
GLIBCXX_3.4.4
GLIBCXX_3.4.5
GLIBCXX_3.4.6
GLIBCXX_3.4.7
GLIBCXX_3.4.8
GLIBCXX_3.4.9
GLIBCXX_3.4.10
GLIBCXX_3.4.11
GLIBCXX_3.4.12
GLIBCXX_3.4.13
GLIBCXX_3.4.14
GLIBCXX_3.4.15
GLIBCXX_3.4.16
GLIBCXX_3.4.17
GLIBCXX_3.4.18
GLIBCXX_3.4.19
GLIBCXX_DEBUG_MESSAGE_LENGTH
```
Just wondering if the little difference between the strings output and the centrifuge output could make any differente. I mean `GLIBCXX_3.4.20 != GLIBCXX_3.4.2`
Bests,
Galo

**Comment by khyox** [comment]
Hi Galo,

Nice to see you around! This is Jose Manuel from I2SysBio. :)

Maybe, you can try ldd and check that everything is in its place. In my case (I compiled it from sources using GCC 6.2.0):
```
 > ldd /software/centrifuge/centrifuge-class
	linux-vdso.so.1 =>  (0x00007fff509fe000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00002b6939c60000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00002b6939e7c000)
	libstdc++.so.6 => /software/gcc-6.2.0/lib64/libstdc++.so.6 (0x00002b693a080000)
	libm.so.6 => /lib64/libm.so.6 (0x00002b693a401000)
	libgcc_s.so.1 => /software/gcc-6.2.0/lib64/libgcc_s.so.1 (0x00002b693a703000)
	libc.so.6 => /lib64/libc.so.6 (0x00002b693a91a000)
	/lib64/ld-linux-x86-64.so.2 (0x00002b6939a3c000)
```
Saludos!
J.Mn.

**Comment by GA-Goig** [comment]
Hi Jose Manuel! Thanks for the advice.
I tried ldd as you pointed but to be honest I don't know what I am actually looking at

```
ldd /data/Software/centrifuge/centrifuge-class 
/data/Software/centrifuge/centrifuge-class: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /data/Software/centrifuge/centrifuge-class)
        linux-vdso.so.1 =>  (0x00007ffe5f787000)
        libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fa404a32000)
        libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fa404729000)
        libm.so.6 => /lib64/libm.so.6 (0x00007fa404426000)
        libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fa404210000)
        libc.so.6 => /lib64/libc.so.6 (0x00007fa403e4f000)
        /lib64/ld-linux-x86-64.so.2 (0x00007fa404c67000)
```
Saludos!

**Comment by khyox** [comment]
I suggest trying with `-v` and looking for missing requirements under `Version information` for centrifuge-class binary itself and under `/lib64/libstdc++.so.6`. In my case I see:
```
	Version information:
	/software/centrifuge/centrifuge-class:
		ld-linux-x86-64.so.2 (GLIBC_2.3) => /lib64/ld-linux-x86-64.so.2
		libdl.so.2 (GLIBC_2.2.5) => /lib64/libdl.so.2
		libgcc_s.so.1 (GCC_3.0) => /software/gcc-6.2.0/lib64/libgcc_s.so.1
		libm.so.6 (GLIBC_2.2.5) => /lib64/libm.so.6
		libstdc++.so.6 (GLIBCXX_3.4.20) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libstdc++.so.6 (CXXABI_1.3.8) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libstdc++.so.6 (GLIBCXX_3.4.11) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libstdc++.so.6 (CXXABI_1.3.9) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libstdc++.so.6 (GLIBCXX_3.4.9) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libstdc++.so.6 (GLIBCXX_3.4.21) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libstdc++.so.6 (CXXABI_1.3) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libstdc++.so.6 (GLIBCXX_3.4) => /software/gcc-6.2.0/lib64/libstdc++.so.6
		libpthread.so.0 (GLIBC_2.2.5) => /lib64/libpthread.so.0
		libpthread.so.0 (GLIBC_2.3.2) => /lib64/libpthread.so.0
		libc.so.6 (GLIBC_2.4) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.14) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.3.4) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.7) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.3) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.2.5) => /lib64/libc.so.6
(...)
	/software/gcc-6.2.0/lib64/libstdc++.so.6:
		ld-linux-x86-64.so.2 (GLIBC_2.3) => /lib64/ld-linux-x86-64.so.2
		libm.so.6 (GLIBC_2.2.5) => /lib64/libm.so.6
		libgcc_s.so.1 (GCC_4.2.0) => /software/gcc-6.2.0/lib64/libgcc_s.so.1
		libgcc_s.so.1 (GCC_3.3) => /software/gcc-6.2.0/lib64/libgcc_s.so.1
		libgcc_s.so.1 (GCC_3.0) => /software/gcc-6.2.0/lib64/libgcc_s.so.1
		libc.so.6 (GLIBC_2.14) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.3) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.17) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.3.2) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.2.5) => /lib64/libc.so.6
```

**Comment by GA-Goig** [comment]
Definetly it can't locate the 3.4.20 version

```
Version information:
        /data/Software/centrifuge/centrifuge-class:
                libgcc_s.so.1 (GCC_3.0) => /lib64/libgcc_s.so.1
                libpthread.so.0 (GLIBC_2.2.5) => /lib64/libpthread.so.0
                libm.so.6 (GLIBC_2.2.5) => /lib64/libm.so.6
                libc.so.6 (GLIBC_2.14) => /lib64/libc.so.6
                libc.so.6 (GLIBC_2.2.5) => /lib64/libc.so.6
                libstdc++.so.6 (GLIBCXX_3.4.11) => /lib64/libstdc++.so.6
                libstdc++.so.6 (CXXABI_1.3) => /lib64/libstdc++.so.6
                **libstdc++.so.6 (GLIBCXX_3.4.20) => not found**
                libstdc++.so.6 (GLIBCXX_3.4.9) => /lib64/libstdc++.so.6
                libstdc++.so.6 (GLIBCXX_3.4) => /lib64/libstdc++.so.6
```
Maybe it is because I have the 3.4.2 instead of 3.4.20 and it is just that zero that is messing things up =S

**Comment by khyox** [comment]
Concretely, under `Version information` for centrifuge-class binary I would look for the `libstdc++.so.6 (GLIBCXX_3.4.21)` entry. If that is pointing to nothing or to your `/lib64/libstdc++.so.6` you have there the issue because as you listed before, it is too old for GLIBCXX_3.4.21. I guess that you must have built that package using a different version of gcc and/or libstdc++. In my case, I use a module system for managing the environment variables. If I purge the modules that I used for compiling the library (and so happens with the environment variables), I reproduce a similar problem:
```
 > /software/centrifuge/centrifuge-class
/software/centrifuge/centrifuge-class: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by /software/centrifuge/centrifuge-class)
(...)
/software/centrifuge/centrifuge-class: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by /software/centrifuge/centrifuge-class)
```

And then, with `ldd -v`:

```
	Version information:
	/software/centrifuge/centrifuge-class:
		ld-linux-x86-64.so.2 (GLIBC_2.3) => /lib64/ld-linux-x86-64.so.2
		libdl.so.2 (GLIBC_2.2.5) => /lib64/libdl.so.2
		libgcc_s.so.1 (GCC_3.0) => /lib64/libgcc_s.so.1
		libm.so.6 (GLIBC_2.2.5) => /lib64/libm.so.6
		libstdc++.so.6 (GLIBCXX_3.4.20) => not found
		libstdc++.so.6 (CXXABI_1.3.8) => not found
		libstdc++.so.6 (GLIBCXX_3.4.11) => /lib64/libstdc++.so.6
		libstdc++.so.6 (CXXABI_1.3.9) => not found
		libstdc++.so.6 (GLIBCXX_3.4.9) => /lib64/libstdc++.so.6
		libstdc++.so.6 (GLIBCXX_3.4.21) => not found
		libstdc++.so.6 (CXXABI_1.3) => /lib64/libstdc++.so.6
		libstdc++.so.6 (GLIBCXX_3.4) => /lib64/libstdc++.so.6
		libpthread.so.0 (GLIBC_2.2.5) => /lib64/libpthread.so.0
		libpthread.so.0 (GLIBC_2.3.2) => /lib64/libpthread.so.0
		libc.so.6 (GLIBC_2.4) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.14) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.3.4) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.7) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.3) => /lib64/libc.so.6
		libc.so.6 (GLIBC_2.2.5) => /lib64/libc.so.6
```

**Comment by GA-Goig** [comment]
Hi again!

Thank you for your feedback. I finally fixed this issue as follows:

I looked for any libstdc*so.6 in my system and checked with `strings` that GLIBCXX_3.4.2 was in it
Then I redownloaded centrifuge with git, and before running Make I set the LD_LIBRARY_PATH
to the path I found this library, which in my case was:
`export LD_LIBRARY_PATH=/data/Software/miniconda3/pkgs/libgcc-5.2.0-0/lib/`

Then I ran make and everything went ok.

Maybe is not the "right" way to do it but, you know, it worked!
